// atem-connection/dist/commands/MixEffects/FadeToBlack/FadeToBlackRateCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractCommand_1 = require("../../AbstractCommand");
const __1 = require("../../..");
class FadeToBlackRateCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'FtbC';
    }
    serialize() {
        const buffer = Buffer.alloc(4);
        buffer.writeUInt8(1, 0);
        buffer.writeUInt8(this.mixEffect, 1);
        buffer.writeUInt8(this.properties.rate, 2);
        return buffer;
    }
}
exports.FadeToBlackRateCommand = FadeToBlackRateCommand;
class FadeToBlackRateUpdateCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'FtbP';
    }
    deserialize(rawCommand) {
        this.mixEffect = __1.Util.parseNumberBetween(rawCommand[0], 0, 3);
        this.properties = {
            rate: rawCommand.readUInt8(1)
        };
    }
    applyToState(state) {
        const mixEffect = state.video.getMe(this.mixEffect);
        mixEffect.fadeToBlack = Object.assign(Object.assign({}, mixEffect.fadeToBlack), { rate: this.properties.rate });
        return `video.ME.${this.mixEffect}.fadeToBlack`;
    }
}
exports.FadeToBlackRateUpdateCommand = FadeToBlackRateUpdateCommand;
