// atem-connection/dist/commands/Macro/MacroPropertiesCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractCommand_1 = require("../AbstractCommand");
const atemUtil_1 = require("../../lib/atemUtil");
class MacroPropertiesCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'MPrp';
    }
    deserialize(rawCommand) {
        this.macroIndexID = rawCommand.readUInt16BE(0);
        const nameLen = rawCommand.readUInt16BE(4);
        const descLen = rawCommand.readUInt16BE(6);
        this.properties = {
            description: '',
            isUsed: Boolean(rawCommand[2] & 1 << 0),
            macroIndex: this.macroIndexID,
            name: ''
        };
        if (descLen > 0) {
            this.properties.description = atemUtil_1.Util.bufToNullTerminatedString(rawCommand, (8 + nameLen), descLen);
        }
        if (nameLen > 0) {
            this.properties.name = atemUtil_1.Util.bufToNullTerminatedString(rawCommand, 8, nameLen);
        }
    }
    applyToState(state) {
        state.macro.macroProperties[this.macroIndexID] = Object.assign(Object.assign({}, state.macro.macroProperties[this.macroIndexID]), this.properties);
        return `macro.macroProperties.${this.macroIndexID}`;
    }
}
exports.MacroPropertiesCommand = MacroPropertiesCommand;
