// atem-connection/dist/commands/MixEffects/PreviewInputCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractCommand_1 = require("../AbstractCommand");
const __1 = require("../..");
class PreviewInputCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'CPvI';
    }
    serialize() {
        const buffer = Buffer.alloc(4);
        buffer.writeUInt8(this.mixEffect, 0);
        buffer.writeUInt16BE(this.properties.source, 2);
        return buffer;
    }
}
exports.PreviewInputCommand = PreviewInputCommand;
class PreviewInputUpdateCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'PrvI';
    }
    deserialize(rawCommand) {
        this.mixEffect = __1.Util.parseNumberBetween(rawCommand[0], 0, 3);
        this.properties = {
            source: rawCommand.readUInt16BE(2)
        };
    }
    applyToState(state) {
        const mixEffect = state.video.getMe(this.mixEffect);
        mixEffect.previewInput = this.properties.source;
        return `video.ME.${this.mixEffect}.previewInput`;
    }
}
exports.PreviewInputUpdateCommand = PreviewInputUpdateCommand;
