// atem-connection/dist/commands/DownstreamKey/DownstreamKeyMaskCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractCommand_1 = require("../AbstractCommand");
class DownstreamKeyMaskCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'CDsM';
    }
    serialize() {
        const buffer = Buffer.alloc(12);
        buffer.writeUInt8(this.flag, 0);
        buffer.writeUInt8(this.downstreamKeyerId, 1);
        buffer.writeUInt8(this.properties.enabled ? 1 : 0, 2);
        buffer.writeInt16BE(this.properties.top, 4);
        buffer.writeInt16BE(this.properties.bottom, 6);
        buffer.writeInt16BE(this.properties.left, 8);
        buffer.writeInt16BE(this.properties.right, 10);
        return buffer;
    }
    updateProps(newProps) {
        this._updateProps(newProps);
    }
}
exports.DownstreamKeyMaskCommand = DownstreamKeyMaskCommand;
DownstreamKeyMaskCommand.MaskFlags = {
    enabled: 1 << 0,
    top: 1 << 1,
    bottom: 1 << 2,
    left: 1 << 3,
    right: 1 << 4
};
