// atem-connection/dist/commands/MixEffects/Transition/TransitionPropertiesCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractCommand_1 = require("../../AbstractCommand");
const __1 = require("../../..");
class TransitionPropertiesCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'CTTp';
    }
    updateProps(newProps) {
        this._updateProps(newProps);
    }
    serialize() {
        const buffer = Buffer.alloc(4);
        buffer.writeUInt8(this.flag, 0);
        buffer.writeUInt8(this.mixEffect, 1);
        buffer.writeUInt8(this.properties.style, 2);
        buffer.writeUInt8(this.properties.selection, 3);
        return buffer;
    }
}
exports.TransitionPropertiesCommand = TransitionPropertiesCommand;
TransitionPropertiesCommand.MaskFlags = {
    style: 1 << 0,
    selection: 1 << 1
};
class TransitionPropertiesUpdateCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'TrSS';
    }
    deserialize(rawCommand) {
        this.mixEffect = __1.Util.parseNumberBetween(rawCommand[0], 0, 3);
        this.properties = {
            style: __1.Util.parseEnum(rawCommand[1], __1.Enums.TransitionStyle),
            selection: rawCommand[2],
            nextStyle: __1.Util.parseEnum(rawCommand[3], __1.Enums.TransitionStyle),
            nextSelection: rawCommand[4]
        };
    }
    applyToState(state) {
        const mixEffect = state.video.getMe(this.mixEffect);
        mixEffect.transitionProperties = Object.assign({}, this.properties);
        return `video.ME.${this.mixEffect}.transitionProperties`;
    }
}
exports.TransitionPropertiesUpdateCommand = TransitionPropertiesUpdateCommand;
