// atem-connection/dist/commands/MixEffects/Key/MixEffectKeyLumaCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractCommand_1 = require("../../AbstractCommand");
const __1 = require("../../..");
class MixEffectKeyLumaCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'CKLm';
    }
    serialize() {
        const buffer = Buffer.alloc(12);
        buffer.writeUInt8(this.flag, 0);
        buffer.writeUInt8(this.mixEffect, 1);
        buffer.writeUInt8(this.upstreamKeyerId, 2);
        buffer.writeUInt8(this.properties.preMultiplied ? 1 : 0, 3);
        buffer.writeUInt16BE(this.properties.clip, 4);
        buffer.writeUInt16BE(this.properties.gain, 6);
        buffer.writeUInt8(this.properties.invert ? 1 : 0, 8);
        return buffer;
    }
}
exports.MixEffectKeyLumaCommand = MixEffectKeyLumaCommand;
MixEffectKeyLumaCommand.MaskFlags = {
    preMultiplied: 1 << 0,
    clip: 1 << 1,
    gain: 1 << 2,
    invert: 1 << 3
};
class MixEffectKeyLumaUpdateCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'KeLm';
    }
    deserialize(rawCommand) {
        this.mixEffect = __1.Util.parseNumberBetween(rawCommand[0], 0, 3);
        this.upstreamKeyerId = __1.Util.parseNumberBetween(rawCommand[1], 0, 3);
        this.properties = {
            preMultiplied: rawCommand[2] === 1,
            clip: __1.Util.parseNumberBetween(rawCommand.readUInt16BE(4), 0, 1000),
            gain: __1.Util.parseNumberBetween(rawCommand.readUInt16BE(6), 0, 1000),
            invert: rawCommand[8] === 1
        };
    }
    applyToState(state) {
        const mixEffect = state.video.getMe(this.mixEffect);
        const upstreamKeyer = mixEffect.getUpstreamKeyer(this.upstreamKeyerId);
        upstreamKeyer.lumaSettings = Object.assign({}, this.properties);
        return `video.ME.${this.mixEffect}.upstreamKeyers.${this.upstreamKeyerId}.lumaSettings`;
    }
}
exports.MixEffectKeyLumaUpdateCommand = MixEffectKeyLumaUpdateCommand;
