// atem-connection/dist/commands/DeviceProfile/productIdentifierCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractCommand_1 = require("../AbstractCommand");
const atemUtil_1 = require("../../lib/atemUtil");
const __1 = require("../..");
class ProductIdentifierCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = '_pin';
    }
    deserialize(rawCommand) {
        this.properties = {
            deviceName: atemUtil_1.Util.bufToNullTerminatedString(rawCommand, 0, 40),
            model: rawCommand[40]
        };
    }
    applyToState(state) {
        state.info.productIdentifier = this.properties.deviceName;
        state.info.model = this.properties.model;
        // Model specific features that aren't specified by the protocol
        switch (state.info.model) {
            case __1.Enums.Model.TwoME:
            case __1.Enums.Model.TwoME4K:
            case __1.Enums.Model.TwoMEBS4K:
            case __1.Enums.Model.Constellation:
            case __1.Enums.Model.Constellation8K:
                state.info.power = [false, false];
                break;
            default:
                state.info.power = [false];
                break;
        }
        return `info`;
    }
}
exports.ProductIdentifierCommand = ProductIdentifierCommand;
