// atem-connection/dist/state/video/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class MixEffect {
    constructor(index) {
        this.transitionProperties = {};
        this.transitionSettings = {};
        this.upstreamKeyers = [];
        this.index = index;
    }
    getUpstreamKeyer(index) {
        if (!this.upstreamKeyers[index]) {
            this.upstreamKeyers[index] = {
                dveSettings: {},
                chromaSettings: {},
                lumaSettings: {},
                patternSettings: {},
                flyKeyframes: [],
                flyProperties: {}
            };
        }
        return this.upstreamKeyers[index];
    }
}
exports.MixEffect = MixEffect;
class SuperSource {
    constructor(index) {
        this.boxes = {};
        this.index = index;
    }
}
exports.SuperSource = SuperSource;
class AtemVideoState {
    constructor() {
        this.ME = {};
        this.downstreamKeyers = {};
        this.auxilliaries = {};
        this.superSources = {};
    }
    getMe(index) {
        if (!this.ME[index]) {
            this.ME[index] = new MixEffect(index);
        }
        return this.ME[index];
    }
    getSuperSource(index) {
        if (!this.superSources[index]) {
            this.superSources[index] = new SuperSource(index);
        }
        return this.superSources[index];
    }
    getDownstreamKeyer(index) {
        if (!this.downstreamKeyers[index]) {
            this.downstreamKeyers[index] = {};
        }
        return this.downstreamKeyers[index];
    }
}
exports.AtemVideoState = AtemVideoState;
