// atem-connection/dist/commands/Media/MediaPlayerSourceCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractCommand_1 = require("../AbstractCommand");
class MediaPlayerSourceCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'MPSS';
    }
    updateProps(newProps) {
        this._updateProps(newProps);
    }
    serialize() {
        const buffer = Buffer.alloc(8);
        buffer.writeUInt8(this.flag, 0);
        buffer.writeUInt8(this.mediaPlayerId, 1);
        buffer.writeUInt8(this.properties.sourceType, 2);
        buffer.writeUInt8(this.properties.stillIndex, 3);
        buffer.writeUInt8(this.properties.clipIndex, 4);
        return buffer;
    }
}
exports.MediaPlayerSourceCommand = MediaPlayerSourceCommand;
MediaPlayerSourceCommand.MaskFlags = {
    sourceType: 1 << 0,
    stillIndex: 1 << 1,
    clipIndex: 1 << 2
};
class MediaPlayerSourceUpdateCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'MPCE';
    }
    deserialize(rawCommand) {
        this.mediaPlayerId = rawCommand[0];
        this.properties = {
            sourceType: rawCommand[1],
            stillIndex: rawCommand[2],
            clipIndex: rawCommand[3]
        };
    }
    applyToState(state) {
        state.media.players[this.mediaPlayerId] = Object.assign(Object.assign({}, state.media.players[this.mediaPlayerId]), this.properties);
        return `media.players.${this.mediaPlayerId}`;
    }
}
exports.MediaPlayerSourceUpdateCommand = MediaPlayerSourceUpdateCommand;
