// atem-connection/dist/commands/Media/MediaPlayerStatusCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractCommand_1 = require("../AbstractCommand");
const atemUtil_1 = require("../../lib/atemUtil");
class MediaPlayerStatusCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'SCPS';
    }
    updateProps(newProps) {
        this._updateProps(newProps);
    }
    serialize() {
        const buffer = Buffer.alloc(8);
        buffer.writeUInt8(this.flag, 0);
        buffer.writeUInt8(this.mediaPlayerId, 1);
        buffer.writeUInt8(this.properties.playing ? 1 : 0, 2);
        buffer.writeUInt8(this.properties.loop ? 1 : 0, 3);
        buffer.writeUInt8(this.properties.atBeginning ? 1 : 0, 4);
        buffer.writeUInt16BE(this.properties.clipFrame, 6);
        return buffer;
    }
}
exports.MediaPlayerStatusCommand = MediaPlayerStatusCommand;
MediaPlayerStatusCommand.MaskFlags = {
    playing: 1 << 0,
    loop: 1 << 1,
    atBeginning: 1 << 2,
    frame: 1 << 3
};
class MediaPlayerStatusUpdateCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'RCPS';
    }
    deserialize(rawCommand) {
        this.mediaPlayerId = atemUtil_1.Util.parseNumberBetween(rawCommand[0], 0, 3);
        this.properties = {
            playing: rawCommand[1] === 1,
            loop: rawCommand[2] === 1,
            atBeginning: rawCommand[3] === 1,
            clipFrame: rawCommand[4] << 8 | (rawCommand[5])
        };
    }
    applyToState(state) {
        state.media.players[this.mediaPlayerId] = Object.assign(Object.assign({}, state.media.players[this.mediaPlayerId]), this.properties);
        return `media.players.${this.mediaPlayerId}`;
    }
}
exports.MediaPlayerStatusUpdateCommand = MediaPlayerStatusUpdateCommand;
