// atem-connection/dist/commands/AuxSourceCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractCommand_1 = require("./AbstractCommand");
class AuxSourceCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'CAuS';
    }
    serialize() {
        const buffer = Buffer.alloc(4);
        buffer.writeUInt8(0x01, 0);
        buffer.writeUInt8(this.auxBus, 1);
        buffer.writeUInt16BE(this.properties.source, 2);
        return buffer;
    }
}
exports.AuxSourceCommand = AuxSourceCommand;
class AuxSourceUpdateCommand extends AbstractCommand_1.default {
    constructor() {
        super(...arguments);
        this.rawName = 'AuxS';
    }
    deserialize(rawCommand) {
        this.auxBus = rawCommand[0];
        this.properties = {
            source: rawCommand.readUInt16BE(2)
        };
    }
    applyToState(state) {
        state.video.auxilliaries[this.auxBus] = this.properties.source;
        return `video.auxilliaries.${this.auxBus}`;
    }
}
exports.AuxSourceUpdateCommand = AuxSourceUpdateCommand;
